File: sdks/dotnet/src/Rrflow.Rrd.Client/ProtocolCodec.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;

namespace Rrflow.Rrd;

internal static class ProtocolCodec
{
    private static readonly JsonSerializerOptions SerializerOptions = new(JsonSerializerDefaults.Web);

    internal static byte[]? EncodeRequest(OperationBinding binding, object? payload)
    {
        if (binding.Endpoint.Method == "GET")
        {
            return null;
        }
        return JsonSerializer.SerializeToUtf8Bytes(new Dictionary<string, object?>
        {
            ["protocol"] = "rrd",
            ["protocol_version"] = 1,
            ["context"] = binding.RequestContext,
            ["resource"] = new Dictionary<string, object?>
            {
                ["segments"] = binding.Resource,
            },
            ["payload"] = payload ?? new Dictionary<string, object?>(),
        }, SerializerOptions);
    }

    internal static JsonElement DecodeResponse(
        int status,
        byte[] encoded,
        IReadOnlyDictionary<string, object?>? requestContext)
    {
        try
        {
            using JsonDocument document = JsonDocument.Parse(encoded);
            JsonElement envelope = document.RootElement;
            RequireFields(
                envelope,
                "protocol",
                "protocol_version",
                "request_id",
                "operation_id",
                "outcome");
            if (envelope.GetProperty("protocol").GetString() != "rrd"
                || envelope.GetProperty("protocol_version").GetInt32() != 1)
            {
                throw new RrdClientException("RRD response protocol differs");
            }
            if (requestContext is not null
                && (envelope.GetProperty("request_id").GetString()
                        != (string)requestContext["request_id"]!
                    || envelope.GetProperty("operation_id").GetString()
                        != (string)requestContext["operation_id"]!))
            {
                throw new RrdClientException("RRD response request/operation identity differs");
            }
            JsonElement outcome = envelope.GetProperty("outcome");
            string outcomeStatus = outcome.GetProperty("status").GetString() ?? string.Empty;
            bool success = status is >= 200 and < 300;
            if (success != (outcomeStatus == "ok"))
            {
                throw new RrdClientException("RRD HTTP status and typed outcome disagree");
            }
            if (outcomeStatus == "error")
            {
                RequireFields(outcome, "status", "error");
                JsonElement error = outcome.GetProperty("error");
                bool hasDetails = error.TryGetProperty("details", out JsonElement detailsElement);
                RequireFields(
                    error,
                    hasDetails
                        ? new[] { "code", "message", "retryable", "details" }
                        : new[] { "code", "message", "retryable" });
                Dictionary<string, string> details = hasDetails
                    ? detailsElement.EnumerateObject().ToDictionary(
                        property => property.Name,
                        property => property.Value.GetString() ?? string.Empty,
                        StringComparer.Ordinal)
                    : new(StringComparer.Ordinal);
                throw new RrdApiException(
                    status,
                    error.GetProperty("code").GetString() ?? string.Empty,
                    error.GetProperty("message").GetString() ?? string.Empty,
                    error.GetProperty("retryable").GetBoolean(),
                    details);
            }
            if (outcomeStatus != "ok")
            {
                throw new RrdClientException("RRD response outcome status is invalid");
            }
            RequireFields(outcome, "status", "payload");
            JsonElement result = outcome.GetProperty("payload");
            if (result.ValueKind != JsonValueKind.Object)
            {
                throw new RrdClientException("RRD success payload must be an object");
            }
            return result.Clone();
        }
        catch (RrdClientException)
        {
            throw;
        }
        catch (Exception error) when (error is JsonException or InvalidOperationException)
        {
            throw new RrdClientException("RRD response envelope is invalid", error);
        }
    }

    private static void RequireFields(JsonElement element, params string[] expected)
    {
        if (element.ValueKind != JsonValueKind.Object)
        {
            throw new RrdClientException("RRD response object is invalid");
        }
        HashSet<string> actual = element.EnumerateObject()
            .Select(property => property.Name)
            .ToHashSet(StringComparer.Ordinal);
        if (!actual.SetEquals(expected))
        {
            throw new RrdClientException("RRD response object fields differ");
        }
    }
}

File: sdks/java/src/main/java/io/rrflow/rrd/ProtocolCodec.java
package io.rrflow.rrd;

import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;

/** Current partial RRD envelope codec; H-04 owns complete generated validation. */
final class ProtocolCodec {
    private static final JsonMapper JSON = JsonMapper.builder().build();

    byte[] encodeRequest(
            Map<String, Object> context, List<ResourceSegment> resource, Object payload) {
        Map<String, Object> envelope = new HashMap<>();
        envelope.put("protocol", "rrd");
        envelope.put("protocol_version", 1);
        envelope.put("context", context);
        envelope.put("resource", Map.of("segments", resource));
        envelope.put("payload", payload == null ? Map.of() : payload);
        try {
            return JSON.writeValueAsBytes(envelope);
        } catch (Exception error) {
            throw new RrdClientException("encode RRD request", error);
        }
    }

    JsonNode decodeResponse(int status, byte[] encoded, Map<String, Object> context) {
        final JsonNode envelope;
        try {
            envelope = JSON.readTree(encoded);
        } catch (Exception error) {
            throw new RrdClientException("RRD response envelope is invalid", error);
        }
        requireExactFields(
                envelope,
                Set.of("protocol", "protocol_version", "request_id", "operation_id", "outcome"));
        if (!envelope.path("protocol").asString().equals("rrd")
                || envelope.path("protocol_version").asInt() != 1) {
            throw new RrdClientException("RRD response protocol differs");
        }
        if (context != null
                && (!envelope.path("request_id").asString().equals(context.get("request_id"))
                        || !envelope
                                .path("operation_id")
                                .asString()
                                .equals(context.get("operation_id")))) {
            throw new RrdClientException("RRD response request/operation identity differs");
        }
        JsonNode outcome = envelope.path("outcome");
        String outcomeStatus = outcome.path("status").asString();
        boolean success = status >= 200 && status < 300;
        if (success != outcomeStatus.equals("ok")) {
            throw new RrdClientException("RRD HTTP status and typed outcome disagree");
        }
        if (outcomeStatus.equals("error")) {
            requireExactFields(outcome, Set.of("status", "error"));
            JsonNode error = outcome.path("error");
            boolean hasDetails = error.has("details");
            requireExactFields(
                    error,
                    hasDetails
                            ? Set.of("code", "message", "retryable", "details")
                            : Set.of("code", "message", "retryable"));
            if (error.path("code").asString().isEmpty()
                    || error.path("message").asString().isEmpty()
                    || !error.path("retryable").isBoolean()
                    || (hasDetails && !error.path("details").isObject())) {
                throw new RrdClientException("RRD error outcome is incomplete");
            }
            Map<String, String> details = new HashMap<>();
            if (hasDetails) {
                error.path("details")
                        .properties()
                        .forEach(entry ->
                                details.put(entry.getKey(), entry.getValue().asString()));
            }
            throw new RrdApiException(
                    status,
                    error.path("code").asString(),
                    error.path("message").asString(),
                    error.path("retryable").asBoolean(),
                    details);
        }
        if (!outcomeStatus.equals("ok")) {
            throw new RrdClientException("RRD response outcome status is invalid");
        }
        requireExactFields(outcome, Set.of("status", "payload"));
        JsonNode result = outcome.path("payload");
        if (!result.isObject()) {
            throw new RrdClientException("RRD success payload must be an object");
        }
        return result;
    }

    private static void requireExactFields(JsonNode node, Set<String> expected) {
        if (node == null || !node.isObject()) {
            throw new RrdClientException("RRD response object is invalid");
        }
        Set<String> actual = new HashSet<>();
        actual.addAll(node.propertyNames());
        if (!actual.equals(expected)) {
            throw new RrdClientException("RRD response object fields differ");
        }
    }
}
